python/pyspark/worker.py

```diff
@@ -59,7 +59,49 @@ def read_command(serializer, file):
 
 def chain(f, g):
     """chain two function together """
-    return lambda x: g(f(x))
+    return lambda *a: g(f(*a))
+
+
+def wrap_udf(f, return_type):
+    return lambda *a: return_type.toInternal(f(*a))
+
+
+def read_single_udf(pickleSer, infile):
+    num_arg = read_int(infile)
+    row_func = None
+    for i in range(read_int(infile)):
+        f, return_type = read_command(pickleSer, infile)
+        if row_func is None:
+            row_func = f
+        else:
+            row_func = chain(row_func, f)
+    # the last returnType will be the return type of UDF
+    return num_arg, wrap_udf(row_func, return_type)
+
+
+def read_udfs(pickleSer, infile):
+    num_udfs = read_int(infile)
+    udfs = []
```

Comment:
It looks like `udfs` holds `(something, something, udf)` triples. Mind adding a line-comment here to say what the first two components of the tuple correspond to?